gubernator/main.py

```diff
@@ -158,24 +167,61 @@ def build_details(build_dir):
         return
     started = json.loads(started)
     finished = json.loads(finished)
+
     failures = []
+    junit_futures = {}
     junit_paths = [f.filename for f in gcs_ls('%s/artifacts' % build_dir)
                    if re.match(r'junit_.*\.xml', os.path.basename(f.filename))]
-    junit_futures = [gcs_async.read(f) for f in junit_paths]
+    
+    fps = []
+    last_len = 0
+    total_len = 0
+    for f in junit_paths:
+        junit_futures[gcs_async.read(f)] = f
+
+
     for future in junit_futures:
         junit = future.get_result()
         if junit is None:
             continue
         failures.extend(parse_junit(junit))
+        total_len = len(fps)
+        last_len = len(failures) - total_len
+        for i in xrange(last_len):
+            fps.append(junit_futures[future])
```

Comment:
I think this would be cleaner if parse_junit included the filename in its output directly. Add a `filename` argument (or keyword argument) to parse_junit, and yield it along with each failure.